File: daily/Arrays/15-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        nums.sort()
        ans = []
        for i in range(len(nums) - 2):
            firstNum = nums[i]
            left, right = i + 1, len(nums) - 1
            if i > 0 and nums[i] == nums[i - 1]:
                continue
            while left < right:
                currentSum = firstNum + nums[left] + nums[right]
                if currentSum > 0:
                    right -= 1
                elif currentSum < 0:
                    left += 1
                else:
                    ans.append([nums[i], nums[left], nums[right]])
                    left += 1
                    right -= 1
                    while nums[left] == nums[left - 1] and left < right:
                        left += 1
        return ans

File: daily/Arrays/15-3sum.go
package daily

import "sort"

func threeSum(nums []int) [][]int {
	sort.Ints(nums)
	var result [][]int

	for i := 0; i < len(nums); i++ {
		if i > 0 && nums[i] == nums[i-1] {
			continue
		}

		left, right := i+1, len(nums)-1
		for left < right {
			current_sum := nums[i] + nums[left] + nums[right]
			if current_sum > 0 {
				right -= 1
			} else if current_sum < 0 {
				left += 1
			} else {
				result = append(result, []int{nums[i], nums[left], nums[right]})
				left += 1
				right -= 1
				for nums[left] == nums[left-1] && left < right {
					left += 1
				}
			}
		}
	}
	return result
}
